tests/unit/s2n_random_test.c

```diff
@@ -894,6 +910,48 @@ int main(int argc, char **argv)
 {
     BEGIN_TEST_NO_INIT();
 
+    /* Feature probe: Negative test */
+    {
+        if (s2n_libcrypto_is_awslc()) {
+#if defined(S2N_LIBCRYPTO_SUPPORTS_ENGINE)
+            FAIL_MSG("Expected ENGINE feature probe to be disabled with AWS-LC");
+#endif
+        }
+    };
+
+    /* Feature probe: Positive test
+     *
+     * Expect ENGINE APIs to be available if `OPENSSL_NO_ENGINE` is not
+     * defined.
+     *
+     * TODO: Test is disabled due to unrelated zlib error on AL2.
+     * https://github.com/aws/s2n-tls/issues/4900
+     *
+     *    {
+     *        if (s2n_libcrypto_is_likely_openssl() && !s2n_is_in_fips_mode()) {
+     *#if !defined(OPENSSL_NO_ENGINE) && !defined(S2N_LIBCRYPTO_SUPPORTS_ENGINE)
+     *            FAIL_MSG("Expected ENGINE feature probe to be enabled");
+     *#endif
+     *        }
+     *    };
+     */
+
+    /* s2n_supports_custom_rand */
+    {
+        if (s2n_supports_custom_rand()) {
+            EXPECT_TRUE(s2n_libcrypto_is_likely_openssl());
+            EXPECT_FALSE(s2n_is_in_fips_mode());
+        }
+
+#if !defined(S2N_LIBCRYPTO_SUPPORTS_ENGINE)
+        EXPECT_FALSE(s2n_supports_custom_rand());
+#endif
+
+#if defined(OPENSSL_NO_ENGINE)
+        EXPECT_FALSE(s2n_supports_custom_rand());
+#endif
```

Comment:
> How/where is OPENSSL_NO_ENGINE defined? Have you tested with a libcrypto with engine support disabled to confirm it's defined where you expect it to be defined?

Yep, I actually included build instructions on how to do this. It the expandable section **Local build instructions**.  `OPENSSL_NO_ENGINE` is lightly documented and the expected behavior when openssl is build with `no-engine` which makes me partial to including it.

Reply:
Based on this [comment](https://github.com/aws/s2n-tls/pull/4705#issuecomment-2301867004), the OPENSSL_NO_ENGINE define comes from `openssl/configuration.h`. Its not clear if that header if being included and hence if this test is executing correctly.

Lesson: we should only rely on s2n defines (s2n-feature probe define or an explicit s2n-cmake define). Relying on an external define, i.e. OPENSSL_NO_ENGINE, requires that we be certain that the correct header is included. Since removing the header can very subtly break code, we should avoid it in general.